subprojects/core/src/main/java/org/gradle/internal/fingerprint/impl/FileCollectionFingerprinterRegistrations.java

```diff
@@ -0,0 +1,122 @@
+
+public class FileCollectionFingerprinterRegistrations {
+    private final Set<FingerprinterRegistration> registrants;
+
+    public FileCollectionFingerprinterRegistrations(
+        StringInterner stringInterner,
+        FileCollectionSnapshotter fileCollectionSnapshotter,
+        ResourceSnapshotterCacheService resourceSnapshotterCacheService,
+        ResourceFilter resourceFilter,
+        ResourceEntryFilter metaInfFilter,
+        Map<String, ResourceEntryFilter> propertiesFileFilters
+        ) {
+
+        CompileClasspathFingerprinter compileClasspathFingerprinter = new DefaultCompileClasspathFingerprinter(resourceSnapshotterCacheService, fileCollectionSnapshotter, stringInterner);
+        this.registrants = concat(
+            withAllDirectorySensitivities(directorySensitivity ->
+                withAllLineEndingSensitivities(lineEndingNormalization -> {
+                    FileSystemLocationSnapshotHasher normalizedContentHasher = normalizedContentHasher(lineEndingNormalization, resourceSnapshotterCacheService);
+                    return Stream.of(
+                        new AbsolutePathFileCollectionFingerprinter(directorySensitivity, lineEndingNormalization, fileCollectionSnapshotter, normalizedContentHasher),
+                        new RelativePathFileCollectionFingerprinter(stringInterner, directorySensitivity, lineEndingNormalization, fileCollectionSnapshotter, normalizedContentHasher),
+                        new NameOnlyFileCollectionFingerprinter(directorySensitivity, lineEndingNormalization, fileCollectionSnapshotter, normalizedContentHasher)
+                    ).map(FingerprinterRegistration::registration);
+                })
+            ),
+            // IgnoredPath and RuntimeClasspath fingperprinters only care about line ending sensitivity
+            withAllLineEndingSensitivities(lineEndingSensitivity -> {
+                    FileSystemLocationSnapshotHasher normalizedContentHasher = normalizedContentHasher(lineEndingSensitivity, resourceSnapshotterCacheService);
+                    return Stream.of(
+                        new IgnoredPathFileCollectionFingerprinter(fileCollectionSnapshotter, lineEndingSensitivity, normalizedContentHasher),
+                        new DefaultClasspathFingerprinter(
+                            resourceSnapshotterCacheService,
+                            fileCollectionSnapshotter,
+                            resourceFilter,
+                            metaInfFilter,
+                            propertiesFileFilters,
+                            stringInterner,
+                            lineEndingSensitivity
+                        )
+                    ).flatMap(fingerprinter -> withAllDirectorySensitivities(directorySensitivity ->
+                        Stream.of(registration(directorySensitivity, fingerprinter.getLineEndingNormalization(), fingerprinter)))
+                    );
```

Comment:
I've refactored this some more so that there is a single stream of line ending sensitivites and directory sensitivities  and we register them all in one place.  I've also pulled the constructors out into separate methods in an attempt to clearly delineate which fingerprinters fall into which category.  Not sure how much this simplifies it, but if you feel like it needs more, let's discuss.